### PR TITLE
feat!: add contacts liveness ffi callbacks

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
- Added the contacts liveness FFI callbacks and destructor methods for the event data
- Expanded the unit test for these callbacks
- Added a cucumber test for these callbacks

Motivation and Context
---
See above.

How Has This Been Tested?
---
- Expanded unit test: `callback_handler_tests`
- Cucumber test: `Scenario: As a client I want to receive contact liveness events`

### CODE REVIEW

File: base_layer/wallet_ffi/src/lib.rs [hunk 1]
```diff
@@ -3313,6 +3316,7 @@ pub unsafe extern "C" fn wallet_create(
     callback_store_and_forward_send_result: unsafe extern "C" fn(c_ulonglong, bool),
     callback_transaction_cancellation: unsafe extern "C" fn(*mut TariCompletedTransaction, u64),
     callback_txo_validation_complete: unsafe extern "C" fn(u64, bool),
+    callback_contacts_liveness_data_updated: unsafe extern "C" fn(*mut TariContactsLivenessData),
```
Comment (philipr-za) [inline base_layer/wallet_ffi/src/lib.rs:3523]: You have not provided the means for an FFI client to destructure the contents of the TariContactsLivenessData into FFI primitives.

Comment (hansieodendaal) [reply base_layer/wallet_ffi/src/lib.rs:3523]: Thanks, well spotted. That was a huge omission.

File: base_layer/wallet_ffi/wallet.h [hunk 2]
```diff
@@ -597,6 +600,21 @@ unsigned long long balance_get_pending_incoming(struct TariBalance *balance, int
 // Gets the available balance from a TariBalance
 unsigned long long balance_get_pending_outgoing(struct TariBalance *balance, int *error_out);
 
+// Gets the public_key from a TariContactsLivenessData
+struct TariPublicKey *liveness_data_get_public_key(struct TariContactsLivenessData *liveness_data, int *error_out);
+
+// Gets the latency from a TariContactsLivenessData
```
Comment (philipr-za) [inline base_layer/wallet_ffi/wallet.h]: What is the unit? Should be documented in the comment

File: base_layer/wallet_ffi/wallet.h [hunk 3]
```diff
@@ -597,6 +600,21 @@ unsigned long long balance_get_pending_incoming(struct TariBalance *balance, int
...
+int liveness_data_get_latency(struct TariContactsLivenessData *liveness_data, int *error_out);
+
+// Gets the last_seen from a TariContactsLivenessData
+char *liveness_data_get_last_seen(struct TariContactsLivenessData *liveness_data, int *error_out);
+
+// Gets the message_type from a TariContactsLivenessData
```
Comment (philipr-za) [inline base_layer/wallet_ffi/wallet.h]: What are the enum options encoded by this int? Shoul be in the docstring comment

File: base_layer/wallet_ffi/wallet.h [hunk 4]
```diff
@@ -597,6 +600,21 @@ unsigned long long balance_get_pending_incoming(struct TariBalance *balance, int
...
+int liveness_data_get_message_type(struct TariContactsLivenessData *liveness_data, int *error_out);
+
+// Gets the online_status from a TariContactsLivenessData
```
Comment (philipr-za) [inline base_layer/wallet_ffi/wallet.h]: What are the options encoded by this Int? Should be documented in the comment.

File: base_layer/wallet_ffi/src/lib.rs [hunk 5]
```diff
@@ -1504,6 +1511,204 @@ pub unsafe extern "C" fn contacts_destroy(contacts: *mut TariContacts) {
...
+
+/// Gets the latency from a TariContactsLivenessData
+///
+/// ## Arguments
+/// `liveness_data` - The pointer to a TariContactsLivenessData
+/// `error_out` - Pointer to an int which will be modified to an error code should one occur, may not be null. Functions
+/// as an out parameter.
+///
+/// ## Returns
+/// `*mut c_int` - Returns a pointer to a c_int if the optional latency data exists, with a value of '-1' if it is None.
```
Comment (philipr-za) [inline base_layer/wallet_ffi/src/lib.rs]: What is the unit this integer signifies?

**Review by philipr-za — COMMENTED** [review]
generally looks fine, mainly some docstring comments.

**Review by philipr-za — APPROVED** [review]
bors r+

**Comment by philipr-za** [comment]
bors retry

**Pull request merged**